flask/launcher/src/main/java/net/corda/flask/launcher/JarCache.java

```diff
@@ -0,0 +1,220 @@
+
+    @Getter
+    private final Path lockFile;
+
+    private final Map<String, Path> extractedLibraries;
+
+
+    public JarCache(String appName) {
+        path = computeCacheDirectory(appName);
+        libDir = path.resolve("lib");
+        lockFile = path.resolve("flask.lock");
+        extractedLibraries = new TreeMap<>();
+    }
+
+    @SneakyThrows
+    public Map<String, Path> extract(Manifest manifest) {
+        byte[] buffer = new byte[0x10000];
+        ClassLoader cl = Launcher.class.getClassLoader();
+        for (Map.Entry<String, Attributes> entry : manifest.getEntries().entrySet()) {
+            String jarEntryName = entry.getKey();
+            Attributes attributes = entry.getValue();
+            if (jarEntryName.startsWith(Flask.Constants.LIBRARIES_FOLDER + '/') && attributes.getValue(Flask.ManifestAttributes.ENTRY_HASH) != null) {
+                String jarName = jarEntryName.substring(jarEntryName.lastIndexOf('/') + 1);
+                String hash = attributes.getValue(Flask.ManifestAttributes.ENTRY_HASH);
+                Path destination = libDir.resolve(hash).resolve(jarName);
+                extractedLibraries.put(hash, destination);
+                if (!Files.exists(destination)) {
+                    Files.createDirectories(destination.getParent());
+                    InputStream is = cl.getResourceAsStream(jarEntryName);
+                    log.debug("Extracting '{}' to '{}'", jarEntryName, destination);
+                    if (is != null) {
+                        try {
+                            try (OutputStream os = new BufferedOutputStream(Files.newOutputStream(destination))) {
+                                while (true) {
+                                    int read = is.read(buffer);
+                                    if (read < 0) break;
+                                    os.write(buffer, 0, read);
+                                }
+                            }
+                        } finally {
+                            is.close();
+                        }
+                    } else {
+                        throw new RuntimeException(String.format("Entry '%s' missing from flask jar", jarEntryName));
+                    }
+                }
+            }
+        }
+        return Collections.unmodifiableMap(extractedLibraries);
+    }
+
+    @SneakyThrows
+    void touchLibraries() {
+        FileTime now = FileTime.from(Instant.now());
+        for(Path jarPath : extractedLibraries.values()) {
+            Files.setLastModifiedTime(jarPath, now);
+        }
+    }
+
+    @SneakyThrows
+    void cleanLibDir() {
+        LockFile lf = LockFile.tryAcquire(lockFile, false);
+        if (lf != null) {
+            log.debug("Starting library cache cleanup");
+            try {
+                FileTime threshold = FileTime.from(Instant.now().minus(Duration.ofDays(7)));
+                log.trace("Removing all files that haven't been touched since {}", threshold);
```

Comment:
yeah, I can make the threshold configurable. so the thing is that I have a cache folder containing all of my required jar files, the cache folder is shared with all the processes that use `flask`, so it can contain old jar files that nobody really uses anymore and could be deleted. Every time a flask Java process terminates, it updates the modification date of all the jar files it has loaded from the cache and before starting it tries to acquire an exclusive lock on the cache folder, if it doesn't manage to acquire the lock it means there is another java process using the same folder, so it gives up, releases the lock and re-acquire it as a shared lock and starts the process, otherwise (should it manage to acquire the exclusive lock), it means it is the only user of the cache and before starting it tries to identify and delete old jar entries (old at the moment is defined as having a "last modification date" older than 7 days)